src/System.Memory/src/System/MemoryExtensions.Fast.cs

```diff
@@ -101,5 +103,120 @@ public static partial class MemoryExtensions
+        /// Attempts to cast a ReadOnlySpan of one primitive type <typeparamref name="TFrom"/> to another primitive type <typeparamref name="TTo"/>.
+        /// </summary>
+        /// <param name="source">The source slice, of type <typeparamref name="TFrom"/>.</param>
+        /// <param name="output">The destination of type <typeparamref name="TTo"/>.</param>
+        /// <remarks>If <typeparamref name="TTo"/> is 8-byte aligned and <paramref name="source"/> points to a valid aligned <typeparamref name="TTo"/> address,
+        /// this will always return true. This is because all C# primitives are aligned against at most 8 bytes, so a pointer that is 8 byte aligned 
+        /// will be aligned to all primitives.</remarks>
+        /// <remarks>If <paramref name="source"/> doesn't point to an address that follows the os-specific alignment rules of <typeparamref name="TFrom"/>, then 
+        /// this will always return false</remarks>
+        /// <remarks>Does not require the memory being spanned over to be fixed.</remarks>
+        /// <returns>True if successful; else False</returns>
+        public static bool TryCast<TFrom, TTo>(this ReadOnlySpan<TFrom> source, out ReadOnlySpan<TTo> output) where TFrom : struct where TTo : struct
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<TFrom>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TFrom));
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<TTo>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TTo));
+
+            if (RuntimeInformation.OSArchitecture == Architecture.X64 ||
+                RuntimeInformation.OSArchitecture == Architecture.X86)
+            {
+                output = MemoryMarshal.Cast<TFrom, TTo>(source);
+                return true;
+            }
+            else
+            {
+                unsafe
+                {
+                    // Test that the source pointer is aligned to either the size of the type or the max alignment size (8), whichever is smaller.
+
+                    // This check will return false in the case where TTo is a struct that is 8 bytes long but only 1/2/4-byte aligned and the source pointer
+                    // is 1/2/4 byte-aligned but not also 8-byte aligned. Unfortunately, to catch that case we would have to walk the 
+                    // struct element sizes to determine the actual struct alignment.
+
+                    // This check also returns false if it's coincidentally aligned to TTo in the case where sizeof(TFrom)<sizeof(TTo). This is
+                    // because the source location can move such that it no longer adheres to the alignment requirement of TTo while still
+                    // following the alignment rules for it's declared/constructed type, TFrom.
+                    void* pointer = Unsafe.AsPointer(ref MemoryMarshal.GetReference(source));
```

Comment:
The GC can move the objects around. For example - on ARM, the `byref` that was 8-byte aligned now can be misaligned 8-byte misaligned a moment later.

Reply:
Interesting. I thought that when GC moves objects, it maintains word alignment.

Reply:
Depends on the GC implementation...

Reply:
>This code is based around two large assumptions:
>The alignment rules for TFrom will be followed if sourceSpan is moved.

This is one of the assumptions made in this implementation.

> for example - on ARM, the byref that was 8-byte aligned now can be 8-byte misaligned a moment later. 

So you're saying that on ARM an array of 8-byte aligned TFrom's can be moved by the GC to a location such that it is no longer 8-byte aligned? Wouldn't that cause any subsequent valid TFrom accesses to fail, since unaligned access is unsupported? How does anything work when on a unaligned-unsupported architecture if the GC is breaking the alignment rules when it moves blocks?

Reply:
> ARM an array of 8-byte aligned TFrom's can be moved by the GC to a location such that it is no longer 8-byte aligned

Right.

> How does anything work when on a unaligned-unsupported architecture if the GC is breaking the alignment rules when it moves blocks?

The GC is not breaking any rules. The GC knows what TFrom is. It will only do it when it is safe (e.g. when TFrom is byte).

Reply:
> It will only do it when it is safe (e.g. when the array type is byte).

Are you talking about the case in which the array type that the GC is aware of has a different alignment than the TFrom that is given to `TryCast`?

I am operating on the assumption that the alignment of TFrom is equivalent to the alignment that the GC is aware of. If they're different then the source Span is already subject to failed unaligned accesses and so our TryCast isn't making anything worse by allowing a cast from an invalid type to another (potentially) invalid type.

So under that assumption, the GC will either:
- maintain the alignment rules of TFrom
- fail whether we allow the cast or not, in which case we might as well allow it.

Reply:
Consider this:

```
// Note this type is guaranteed to be 4-byte aligned. There is nothing forcing it to be 8-byte aligned.
struct Bundle {
    int x,y;
}

Bundle[] a = new Bundle[10]; // Let's assume that the GC will happen to allocate the array on 8-byte boundary
Span<Bundle> s = new Span<Bundle>(a);
bool success = s.TryCast<Bundle,double>(out Span<double> sd); // This will succeed
//GC moves `a` to a different place that is not 8-byte aligned anymore ...
double d = sd[5]; // Misaligned access fault
```

Reply:
I see, thanks for the example. We will return the wrong result in the case where TFrom is a struct in which all of these conditions are met:
1. ```Unsafe.SizeOf<TFrom> != actualTFromAlignment```
2. ```actualTFromAlignment < Unsafe.SizeOf<TTo>```
3. ```addressof(source) % Unsafe.SizeOf<TTo> == 0```

This is essentially the reverse of the issue I commented about above:
```
 // This check will return false in the case where TTo is a struct that is 8 bytes long but only 1/2/4-byte aligned and the source pointer
// is 1/2/4 byte-aligned but not also 8-byte aligned. Unfortunately, to catch that case we would have to walk the 
// struct element sizes to determine the actual struct alignment.
```

Except in this case we return `true` incorrectly while in the other case we return `false` incorrectly, which is of course a more severe error.

Both are the result of the difference between SizeOf<T> and SizeOfLargestElement<T>. I'm not aware of a way to get the latter without reflection.

Reply:
How about we special-case when TFrom is a struct to assume it is only byte-aligned? We can leave in the logic using SizeOf for non-struct types in which we know that ```Max(8, SizeOf<T>) == AlignmentOf<T>```

Reply:
How do you feel about:

```
int sizeOfTFrom = typeof(TFrom).IsPrimitive ? Math.Min(8, Unsafe.SizeOf<TFrom>()) : 1;
```

Without being able to look into the struct to determine the real alignment, we can't really assume it follows any alignment but byte alignment, which means TryCast will pretty much always return false when TFrom is a struct.

Reply:
It will make the API slower for the common use case of casting span of bytes to span of chars; or span of ints.

I think it is a good thing for the portable version of the API to be more conservative. It will tech folks to write proper fallbacks for this.